src/main/java/hudson/remoting/PingThread.java

```diff
@@ -103,13 +103,13 @@ public void run() {
     private void ping() throws IOException, InterruptedException {
         Future<?> f = channel.callAsync(new Ping());
         long start = System.currentTimeMillis();
-        long end = start +timeout;
 
-        long remaining;
+        long end = System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(timeout);
+        long remaining = end - System.nanoTime();
+
         do {
-            remaining = end-System.currentTimeMillis();
             try {
-                f.get(Math.max(0,remaining),MILLISECONDS);
+                f.get(Math.max(TimeUnit.MILLISECONDS.toNanos(10),remaining),TimeUnit.NANOSECONDS);
```

Comment:
There is very little point waiting 0ns and expecting anything different to happen as we are waiting on remote calls it seemed silly to wait anything much smaller.